content/en/blog/_posts/non-root-containers-and-devices.md

```diff
@@ -0,0 +1,196 @@
+---
+layout: blog
+title: 'Non-root Containers And Devices'
+date: 2021-09-YY
+slug: non-root-containers-and-devices
+---
+
+**Author:** Mikko Ylinen (Intel)
+
+The user/group ID related security settings in Pod's `securityContext` trigger a problem when users want to
+deploy containers that use accelerator devices (via [Kubernetes Device Plugins](docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)) on Linux. In this blog
+post we talk about the problem and describe the work that was made so far to address it. It's not meant to be a long story about getting the ["k/k issue"](https://github.com/kubernetes/kubernetes/issues/92211) fixed but to raise the awaress of it and important device use-cases too. This is needed as we work on new related features such as support for user namespaces.
+
+## Why non-root containers can't use devices and why it matters?
+One of the key security principles for running containers in Kubernetes is the
+principle of least privilege. The Pod/container `securityContext` specifies the config
+knobs to set, e.g., Linux* capabilities, MAC policies, and user/group ID values to achieve this.
+
+Furthermore, the cluster admins are supported with tools like [Pod Security Policy (deprecated)](docs/concepts/policy/pod-security-policy/) or
+[Pod Security Admission](docs/concepts/security/pod-security-admission/) to enforce the desired security settings for pods that are being deployed in
+the cluster. These settings could, for instance, require that containers must be `runAsNonRoot` or
+that they are forbidden from running with root's group ID in `runAsGroup` or `supplementalGroups`.
+
+In Kubernetes*, kubelet builds the list of [`Device`](https://pkg.go.dev/k8s.io/cri-api@v0.22.1/pkg/apis/runtime/v1#Device)s to be made available to a container
+(based on inputs from the Device Plugins) and the list is included in the CreateContainer CRI message
+sent to the CRI container runtime. Each `Device` contains little information: host/container device
+paths and the desired devices cgroups permissions.
+
+The [OCI Runtime Spec for Linux Container Configuration](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md)
+expects that in addition to the devices cgroup fields, more detailed information about the devices
+must be provided:
+
+```json
```

Comment:
I expect this will syntax highlight better if you tell the static-site generator a white lie and say it's YAML. (All JSON documents are valid YAML).

Reply:
done.